logstash-core/src/main/java/org/logstash/secret/store/backend/JavaKeyStore.java

```diff
@@ -0,0 +1,227 @@
+            e.printStackTrace();
+        } finally {
+            readLock.unlock();
+        }
+        return identifiers;
+    }
+
+    @Override
+    public void persistSecret(SecretIdentifier identifier, byte[] secret) {
+        try {
+            writeLock.lock();
+            SecretKeyFactory factory = SecretKeyFactory.getInstance("PBE");
+            //PBEKey requires an ascii password, so base64 encode it
+            byte[] base64 = Base64.getEncoder().encode(secret);
+            PBEKeySpec passwordBasedKeySpec = new PBEKeySpec(asciiBytesToChar(base64));
+            SecretKey secretKey = factory.generateSecret(passwordBasedKeySpec);
+            keyStore.setEntry(identifier.toExternalForm(), new KeyStore.SecretKeyEntry(secretKey), protectionParameter);
+            try (final OutputStream os = Files.newOutputStream(keyStorePath)) {
+                keyStore.store(os, keyStorePass);
+            } finally {
+                passwordBasedKeySpec.clearPassword();
+                clearSecret(secret);
+            }
+        } catch (Exception e) {
+            throw new SecretStoreException.PersistException(identifier, e);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void purgeSecret(SecretIdentifier identifier) {
+        try {
+            writeLock.lock();
+            try (final InputStream is = Files.newInputStream(keyStorePath)) {
+                keyStore.load(is, keyStorePass);
+                keyStore.deleteEntry(identifier.toExternalForm());
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+            try (final OutputStream os = Files.newOutputStream(keyStorePath)) {
+                keyStore.store(os, keyStorePass);
+            }
+        } catch (Exception e) {
+            throw new SecretStoreException.PurgeException(identifier, e);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public byte[] retrieveSecret(SecretIdentifier identifier) {
+        if (identifier != null && identifier.getKey() != null && !identifier.getKey().isEmpty()) {
+            try {
+                readLock.lock();
+                try (final InputStream is = Files.newInputStream(keyStorePath)) {
+                    keyStore.load(is, keyStorePass);
+                    SecretKeyFactory factory = SecretKeyFactory.getInstance("PBE");
+                    KeyStore.SecretKeyEntry secretKeyEntry = (KeyStore.SecretKeyEntry) keyStore.getEntry(identifier.toExternalForm(), protectionParameter);
+                    //not found
+                    if (secretKeyEntry == null) {
+                        return null;
+                    }
+                    PBEKeySpec passwordBasedKeySpec = (PBEKeySpec) factory.getKeySpec(secretKeyEntry.getSecretKey(), PBEKeySpec.class);
+                    //base64 encoded char[]
+                    char[] base64secret = passwordBasedKeySpec.getPassword().clone();
```

Comment:
PBEKeySpec.getPassword() docs say "Returns a copy of the password." -- I don't think the .clone() is necessary.

Reply:
will do. thanks.